### PR TITLE
Section5 assignment

### PR DESCRIPTION
**Pull request by ChristopherGS**
Current status:

- Model trains
- Prediction test runs
- Package builds

TODO
- Check assumptions on preprocessing and test file generation
- Run package in another application

### CODE REVIEW

File: assignment-section-05/classification_model/processing/validation.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from classification_model.config.core import config
+from classification_model.processing.data_manager import pre_pipeline_preparation
+
+
+def validate_inputs(*, input_data: pd.DataFrame) -> Tuple[pd.DataFrame, Optional[dict]]:
+    """Check model inputs for unprocessable values."""
+
+    pre_processed = pre_pipeline_preparation(dataframe=input_data)
+    validated_data = pre_processed[config.model_config.features].copy()
+    # TODO: check if we want to drop NA
```
Comment (ChristopherGS) [inline assignment-section-05/classification_model/processing/validation.py:16]: @solegalli do we want to drop NA values? If so, which ones? If we do a blanket drop, almost every row is lost.

File: assignment-section-05/generate_test_data.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+from pathlib import Path
+
+from classification_model.config.core import config, DATASET_DIR
+from classification_model.processing.data_manager import _load_raw_dataset
+
+
+def generate_data_data() -> None:
```
Comment (ChristopherGS) [inline assignment-section-05/generate_test_data.py:7]: @solegalli I downloaded the data as `raw.csv`, put it in the datasets directory, and then ran this script to generate a `test.csv` for use in the testing. My assumption is that we want to keep some part for testing (or can we assume train_test_split always gives the same rows for testing?)

File: assignment-section-05/classification_model/config.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+# Package Overview
+package_name: regression_model
```
Comment (solegalli) [inline assignment-section-05/classification_model/config.yml:2]: classification_model? or titanic_model?

File: assignment-section-05/classification_model/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+# to persist the model and the scaler
```
Comment (solegalli) [inline assignment-section-05/classification_model/pipeline.py:1]: this comment? probably legacy. Pls remove

File: assignment-section-05/classification_model/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+from feature_engine.encoding import RareLabelEncoder, OneHotEncoder
+
+# for imputation
+from feature_engine.imputation import (
+    CategoricalImputer,
+    AddMissingIndicator,
+    MeanMedianImputer,
+)
+
+# for the preprocessors
```
Comment (solegalli) [inline assignment-section-05/classification_model/pipeline.py:12]: this comment? probably remove

File: assignment-section-05/classification_model/predict.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    input_data: t.Union[pd.DataFrame, dict],
+) -> dict:
+    """Make a prediction using a saved model pipeline."""
+
+    data = pd.DataFrame(input_data)
+    validated_data, errors = validate_inputs(input_data=data)
+    results = {"predictions": None, "version": _version, "errors": errors}
+
+    if not errors:
+        predictions = _price_pipe.predict(
```
Comment (solegalli) [inline assignment-section-05/classification_model/predict.py:25]: price_pipe? probably titanic_pipe?

File: assignment-section-05/classification_model/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+from sklearn.pipeline import Pipeline
+
+# feature scaling
+from sklearn.preprocessing import StandardScaler
+
+from classification_model.config.core import config
+from classification_model.processing.features import ExtractLetterTransformer
+
+
+price_pipe = Pipeline(
```
Comment (solegalli) [inline assignment-section-05/classification_model/pipeline.py:26]: titanic_pipe

File: assignment-section-05/classification_model/predict.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+import pandas as pd
+
+from classification_model import __version__ as _version
+from classification_model.config.core import config
+from classification_model.processing.data_manager import load_pipeline
+from classification_model.processing.validation import validate_inputs
+
+pipeline_file_name = f"{config.app_config.pipeline_save_file}{_version}.pkl"
+_price_pipe = load_pipeline(file_name=pipeline_file_name)
```
Comment (solegalli) [inline assignment-section-05/classification_model/predict.py:11]: titanic pipe?

File: .gitignore [hunk 9]
```diff
@@ -112,6 +112,7 @@ packages/regression_model/regression_model/datasets/*.zip
 packages/regression_model/regression_model/datasets/*.txt
 train.csv
 test.csv
+raw.csv
```
Comment (solegalli) [inline .gitignore:115]: Do we need to store it locally? We can fetch from url online

File: assignment-section-05/classification_model/config.yml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+# Data Files
+raw_data_file: raw.csv
+training_data_file: train.csv
+test_data_file: test.csv
+
+# Variables
+# The variable we are attempting to predict (sale price)
```
Comment (solegalli) [inline assignment-section-05/classification_model/config.yml:10]: comment: survival instead of sale price

File: assignment-section-05/classification_model/processing/data_manager.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    data["age"] = data["age"].astype("float")
+
+    # drop unnecessary variables
+    data.drop(labels=config.model_config.unused_fields, axis=1, inplace=True)
+
+    return data
+
+
+def _load_raw_dataset(*, file_name: str) -> pd.DataFrame:
+    dataframe = pd.read_csv(Path(f"{DATASET_DIR}/{file_name}"))
```
Comment (solegalli) [inline assignment-section-05/classification_model/processing/data_manager.py:58]: I did not see the url/link to the online data anywhere. Is this because you ask students somewhere to download the data and store it as raw.csv?

Or maybe I missed the online link?

If you want to start by loading raw.csv, you need to make crystal clear that they need to download the data from the url and save it as raw.csv at the very beginning of the assignment. Otherwise, the students will try to pick it up from the URL as I do in the notebook.

File: assignment-section-05/classification_model/config.yml [hunk 12]
```diff
@@ -0,0 +1,51 @@
+# Package Overview
+package_name: regression_model
+
+# Data Files
+raw_data_file: raw.csv
```
Comment (solegalli) [inline assignment-section-05/classification_model/config.yml:5]: where does this file come from? 

should we not include in the config the url to fetch the data instead ?

File: assignment-section-05/classification_model/processing/validation.py (hunk 1 above)
Comment (solegalli) [reply assignment-section-05/classification_model/processing/validation.py:16]: I think we handle NA in the pipeline. I don't think we need to drop extra NA in this dataset. Alternatively, let's do that just in the variables that require imputation.

File: assignment-section-05/classification_model/processing/validation.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    name: Optional[str]
+    sex: Optional[str]
+    age: Optional[int]
+    sibsp: Optional[int]
+    parch: Optional[int]
+    ticket: Optional[int]
+    fare: Optional[float]
+    cabin: Optional[str]
+    embarked: Optional[str]
+    boat: Optional[Union[str, int]]
```
Comment (solegalli) [inline assignment-section-05/classification_model/processing/validation.py:41]: do we need to validate boat and body? we are not using these as input features.

File: assignment-section-05/classification_model/train_pipeline.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
+from sklearn.model_selection import train_test_split
+
+from config.core import config
+from pipeline import price_pipe
```
Comment (solegalli) [inline assignment-section-05/classification_model/train_pipeline.py:4]: titanic_pipe?

File: assignment-section-05/classification_model/train_pipeline.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    data = load_dataset(file_name=config.app_config.training_data_file)
+
+    # divide train and test
+    X_train, X_test, y_train, y_test = train_test_split(
+        data[config.model_config.features],  # predictors
+        data[config.model_config.target],
+        test_size=config.model_config.test_size,
+        # we are setting the random seed here
+        # for reproducibility
+        random_state=config.model_config.random_state,
```
Comment (solegalli) [inline assignment-section-05/classification_model/train_pipeline.py:21]: supposedly, because we set the random_state, it should always return the same test/train split.

File: assignment-section-05/classification_model/train_pipeline.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    )
+
+    # fit model
+    price_pipe.fit(X_train, y_train)
```
Comment (solegalli) [inline assignment-section-05/classification_model/train_pipeline.py:25]: titanic?

File: assignment-section-05/generate_test_data.py (hunk 2 above)
Comment (solegalli) [reply assignment-section-05/generate_test_data.py:7]: Students won't know where raw.csv comes from. And they would probably try to pick it up from the url directly when refactoring the code. So if you want to proceed with raw.csv, you will need to tell them to download it and save it somewhere early on in the assignment.

if we set the `random_state` in `train_test_split,` supposedly, it should always return the same train/test split.

File: assignment-section-05/generate_test_data.py [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    """
+    Generate train.csv and test.csv files.
+
+    The titanic dataset does not come split into train and test files, so we
+    do this ourselves. Note, this requires you to have downloaded the original
+    datafile from here: https://www.openml.org/data/get_csv/16826755/phpMYEkMl
+    and renamed it to match your config raw_data_file name.
```
Comment (solegalli) [inline assignment-section-05/generate_test_data.py:14]: the instruction to download the dataset is too late here. By now, they would have completely refactored the code. This needs to be done before they start the assignment.

File: assignment-section-05/generate_test_data.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    """
+    data = _load_raw_dataset(file_name=config.app_config.raw_data_file)
+    df_len = len(data)
+    row_split_index = round(df_len*(1-config.model_config.test_size))
```
Comment (solegalli) [inline assignment-section-05/generate_test_data.py:18]: the split done by `train_test_split` is at random. Here you are doing top/bottom as train test, so it is not the same.

Not sure what is the intention of this function and I don't think students will intuitively generate something like this.

If this is needed, it needs to be made clear before they start the assignment. Probably provide a pre-drafted template.

File: assignment-section-05/generate_test_data.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
+from pathlib import Path
```
Comment (solegalli) [inline assignment-section-05/generate_test_data.py:1]: this file is new respect to the house price, isn't it?

I doubt very much that students will imagine doing anything like what is going on here. I would not add extra files.

File: assignment-section-05/mypy.ini [hunk 20]
```diff
@@ -0,0 +1,14 @@
+[mypy]
+# warn_unreachable = True
+warn_unused_ignores = True
+follow_imports = skip
+show_error_context = True
+warn_incomplete_stub = True
+ignore_missing_imports = True
+check_untyped_defs = True
+cache_dir = /dev/null
+# Cannot enable this one as we still allow defining functions without any types.
```
Comment (solegalli) [inline assignment-section-05/mypy.ini:10]: ?

File: assignment-section-05/setup.py [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+DESCRIPTION = "Example Titanic dataset classification model package from Train In Data."
+URL = "https://github.com/trainindata/deploying-machine-learning-models"
+EMAIL = "christopher.samiullah@protonmail.com"
+AUTHOR = "ChristopherGS"
+REQUIRES_PYTHON = ">=3.6.0"
+
+
+# The rest you shouldn't have to touch too much :)
+# ------------------------------------------------
+# Except, perhaps the License and Trove Classifiers!
```
Comment (solegalli) [inline assignment-section-05/setup.py:19]: what is a Trove Classifier?

File: assignment-section-05/tests/test_prediction.py [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+    # When
+    result = make_prediction(input_data=sample_input_data)
+
+    # Then
+    predictions = result.get("predictions")
+    assert isinstance(predictions, np.ndarray)
+    assert isinstance(predictions[0], np.int64)
+    assert result.get("errors") is None
+    assert len(predictions) == expected_no_predictions
+    assert list(predictions)[:10] == first_ten_expected_predictions
```
Comment (solegalli) [inline assignment-section-05/tests/test_prediction.py:19]: we should probably corroborate the actual value of the predictions. We don't do this in the house price either, do we?

**Review by solegalli — COMMENTED** [review]
Hi Chris,

Thank you so much for the code for the assignment. Great work. 

I left some comments throughout and I've got 2 suggestions. 

The first one regarding the source data. If you decide to go for raw.csv, at the beginning of the assignment, you probably need to have a snippet of code showing students where to get the data from and how to save it. Alternatively, we can load from the URL.

The second one is regarding the generate test data file. We do not have such thing in the house price, do we? Thus, I would either, not include it, or, provide it as a template or even already pre-filled, and make clear what is the use of the file before the assignment. Otherwise, I doubt very much anyone would come up with something like this.

The rest are minor comments

Thank you!

File: assignment-section-05/classification_model/processing/validation.py (hunk 1 above)
Comment (ChristopherGSA) [reply assignment-section-05/classification_model/processing/validation.py:16]: OK, skipping NA drop here.

File: assignment-section-05/generate_test_data.py (hunk 2 above)
Comment (ChristopherGSA) [reply assignment-section-05/generate_test_data.py:7]: Agree this is too complex. I've updated the test code to just use `train_test_split` now. I'll update the lecture video to explain downloading the code and saving a single file: raw.csv

Comment (ChristopherGSA) [reply assignment-section-05/generate_test_data.py:7]: I've removed the generate test data script

File: assignment-section-05/classification_model/config.yml (hunk 3 above)
Comment (ChristopherGSA) [reply assignment-section-05/classification_model/config.yml:2]: Because we called the other model `regression_model` I felt this was more consistent.

File: assignment-section-05/classification_model/pipeline.py (hunk 4 above)
Comment (ChristopherGSA) [reply assignment-section-05/classification_model/pipeline.py:1]: removed

File: assignment-section-05/classification_model/pipeline.py (hunk 5 above)
Comment (ChristopherGSA) [reply assignment-section-05/classification_model/pipeline.py:12]: removed (and one or two other v. obvious ones)

File: assignment-section-05/classification_model/pipeline.py (hunk 7 above)
Comment (ChristopherGSA) [reply assignment-section-05/classification_model/pipeline.py:26]: good spot, updated everywhere

File: assignment-section-05/classification_model/predict.py (hunk 6 above)
Comment (ChristopherGSA) [reply assignment-section-05/classification_model/predict.py:25]: updated

File: assignment-section-05/generate_test_data.py (hunk 18 above)
Comment (ChristopherGSA) [reply assignment-section-05/generate_test_data.py:18]: deleted file

File: assignment-section-05/mypy.ini (hunk 20 above)
Comment (ChristopherGSA) [reply assignment-section-05/mypy.ini]: simplified

File: assignment-section-05/setup.py (hunk 21 above)
Comment (ChristopherGSA) [reply assignment-section-05/setup.py:19]: Added an explanation line

File: assignment-section-05/tests/test_prediction.py (hunk 22 above)
Comment (ChristopherGSA) [reply assignment-section-05/tests/test_prediction.py]: Just had a look - predictions appear to be really off. Any idea why the model is so inaccurate?

Comment (ChristopherGSA) [reply assignment-section-05/tests/test_prediction.py]: 0.71 accuracy

File: assignment-section-05/tests/test_prediction.py [hunk 23]
```diff
@@ -21,4 +22,7 @@ def test_make_prediction(sample_input_data, first_twenty_expected_predictions):
     assert result.get("errors") is None
     assert len(predictions) == expected_no_predictions
     _predictions = list(predictions)
-    assert _predictions[:20] == first_twenty_expected_predictions
+    y_true = sample_input_data['survived']
+    accuracy = accuracy_score(_predictions, y_true)
+    assert accuracy > 0.7
```
Comment (ChristopherGSA) [inline assignment-section-05/tests/test_prediction.py:27]: @solegalli presumably we want to aim for higher than this?

File: assignment-section-05/classification_model/config.yml (hunk 3 above)
Comment (solegalli) [reply assignment-section-05/classification_model/config.yml:2]: since this is a classification task, I would anticipate that most students would change the name (regression is normally used for continuous targets). But I guess it is not so important. If it is a small tweak, change it. Otherwise leave it.

File: assignment-section-05/tests/test_prediction.py (hunk 22 above)
Comment (solegalli) [reply assignment-section-05/tests/test_prediction.py]: Most likely because we are using logistic regression. We should use trees for classification.

File: assignment-section-05/tests/test_prediction.py (hunk 23 above)
Comment (solegalli) [reply assignment-section-05/tests/test_prediction.py:27]: If you were making a real model yes. But in the notebook we also have 0.71 and 0.72. So I would not bother too much. The idea was to introduce a variety of engineering steps and I think I used a logistic regression because I wanted to use a scaler to make the pipe a bit more complicated.  Trees do not need scaling of variables.

File: assignment-section-05/classification_model/config.yml (hunk 10 above)
Comment (solegalli) [reply assignment-section-05/classification_model/config.yml:10]: this comment still needs updating
